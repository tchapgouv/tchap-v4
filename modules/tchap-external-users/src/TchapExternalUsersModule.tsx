/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
