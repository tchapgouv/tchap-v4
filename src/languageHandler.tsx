/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
