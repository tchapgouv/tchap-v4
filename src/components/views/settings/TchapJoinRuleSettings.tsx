/* eslint-disable max-len */
/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { ReactNode } from "react";
import { IJoinRuleEventContent, JoinRule, RestrictedAllowType } from "matrix-js-sdk/src/@types/partials";
import { Room } from "matrix-js-sdk/src/models/room";
import { EventType } from "matrix-js-sdk/src/@types/event";
import StyledRadioGroup, { IDefinition } from "matrix-react-sdk/src/components/views/elements/StyledRadioGroup";
import { _t } from "matrix-react-sdk/src/languageHandler";
import AccessibleButton from "matrix-react-sdk/src/components/views/elements/AccessibleButton";
import SpaceStore from "matrix-react-sdk/src/stores/spaces/SpaceStore";
import RoomAvatar from "matrix-react-sdk/src/components/views/avatars/RoomAvatar";
import { MatrixClientPeg } from "matrix-react-sdk/src/MatrixClientPeg";
import Modal from "matrix-react-sdk/src/Modal";
import ManageRestrictedJoinRuleDialog from "matrix-react-sdk/src/components/views/dialogs/ManageRestrictedJoinRuleDialog";
import RoomUpgradeWarningDialog, {
    IFinishedOpts,
} from "matrix-react-sdk/src/components/views/dialogs/RoomUpgradeWarningDialog";
import { upgradeRoom } from "matrix-react-sdk/src/utils/RoomUpgrade";
import { arrayHasDiff } from "matrix-react-sdk/src/utils/arrays";
import { useLocalEcho } from "matrix-react-sdk/src/hooks/useLocalEcho";
import dis from "matrix-react-sdk/src/dispatcher/dispatcher";
import { ROOM_SECURITY_TAB } from "matrix-react-sdk/src/components/views/dialogs/RoomSettingsDialog";
import { Action } from "matrix-react-sdk/src/dispatcher/actions";
import { ViewRoomPayload } from "matrix-react-sdk/src/dispatcher/payloads/ViewRoomPayload";
import { doesRoomVersionSupport, PreferredRoomVersions } from "matrix-react-sdk/src/utils/PreferredRoomVersions";
import LabelledToggleSwitch from "matrix-react-sdk/src/components/views/elements/LabelledToggleSwitch";
import QuestionDialog from "matrix-react-sdk/src/components/views/dialogs/QuestionDialog";

import TchapUIFeature from "../../../tchap/util/TchapUIFeature";
import { TchapRoomAccessRule, TchapIAccessRuleEventContent, TchapRoomAccessRulesEventId } from "../../../tchap/@types/tchap";

interface IProps {
    room: Room;
    promptUpgrade?: boolean;
    closeSettingsFn(): void;
    onError(error: Error): void;
    beforeChange?(joinRule: JoinRule): Promise<boolean>; // if returns false then aborts the change
    aliasWarning?: ReactNode;
}

const JoinRuleSettings = ({ room, promptUpgrade, aliasWarning, onError, beforeChange, closeSettingsFn }: IProps) => {
    const cli = room.client;

    const roomSupportsRestricted = doesRoomVersionSupport(room.getVersion(), PreferredRoomVersions.RestrictedRooms);
    const preferredRestrictionVersion =
        !roomSupportsRestricted && promptUpgrade ? PreferredRoomVersions.RestrictedRooms : undefined;

    const disabled = !room.currentState.mayClientSendStateEvent(EventType.RoomJoinRules, cli);

    const [content, setContent] = useLocalEcho<IJoinRuleEventContent>(
        () => room.currentState.getStateEvents(EventType.RoomJoinRules, "")?.getContent(),
        (content) => cli.sendStateEvent(room.roomId, EventType.RoomJoinRules, content, ""),
        onError,
    );
    const { join_rule: joinRule = JoinRule.Invite } = content || {};
    const restrictedAllowRoomIds =
        joinRule === JoinRule.Restricted
            ? content.allow?.filter((o) => o.type === RestrictedAllowType.RoomMembership).map((o) => o.room_id)
            : undefined;

    const [contentTchapAccessRule, setTchapAccessRule] = useLocalEcho<TchapIAccessRuleEventContent>(
        () => room.currentState.getStateEvents(TchapRoomAccessRulesEventId, "")?.getContent(),
        (content) => cli.sendStateEvent(room.roomId, TchapRoomAccessRulesEventId, content, ""),
        onError,
    );
    const { rule: accessRule = undefined } = contentTchapAccessRule || {};

    const editRestrictedRoomIds = async (): Promise<string[] | undefined> => {
        let selected = restrictedAllowRoomIds;
        if (!selected?.length && SpaceStore.instance.activeSpaceRoom) {
            selected = [SpaceStore.instance.activeSpaceRoom.roomId];
        }

        const matrixClient = MatrixClientPeg.get();
        const { finished } = Modal.createDialog(
            ManageRestrictedJoinRuleDialog,
            {
                matrixClient,
                room,
                selected,
            },
            "mx_ManageRestrictedJoinRuleDialog_wrapper",
        );

        const [roomIds] = await finished;
        return roomIds;
    };

    /* code from element web
    const definitions: IDefinition<JoinRule>[] = [{
        value: JoinRule.Invite,
        label: _t("Private (invite only)"),
        description: _t("Only invited people can join."),
        checked: joinRule === JoinRule.Invite || (joinRule === JoinRule.Restricted && !restrictedAllowRoomIds?.length),
    }, {
        value: JoinRule.Public,
        label: _t("Public"),
        description: <>
            { _t("Anyone can find and join.") }
            { aliasWarning }
        </>,
    }]; */

    // :TCHAP: we do not permit to change the type of room, thus display only one option
    const definitions: IDefinition<JoinRule>[] = [];

    // :TCHAP: do we need to add the following condition as well (joinRule === JoinRule.Restricted && !restrictedAllowRoomIds?.length)?
    if (joinRule === JoinRule.Invite) {
        let privateRoomDescription = <div>{_t("Only invited people can join.")}</div>;
        // :TCHAP: We could add functions in 'TchapUtils' to determine the type of room and rely on this logic to display components as we did in Android :
        // :TCHAP: https://github.com/tchapgouv/tchap-android/blob/develop/vector/src/main/java/fr/gouv/tchap/core/utils/RoomUtils.kt#L31
        if (accessRule) {
            const openedToExternalUsers = accessRule === TchapRoomAccessRule.Unrestricted;
            const onExternalAccessChange = async () => {
                Modal.createDialog(QuestionDialog, {
                    title: _t("Allow external users to join this room"),
                    description:
                        _t("This action is irreversible.") +
                        " " +
                        _t("Are you sure you want to allow the externals to join this room ?"),
                    button: _t("OK"),
                    onFinished: (confirmed) => {
                        if (!confirmed) return;
                        setTchapAccessRule({ rule: TchapRoomAccessRule.Unrestricted });
                    },
                });
            };
            privateRoomDescription = (
                <div>
                    <div>{_t("Only invited people can join.")}</div>
                    <span>
                        <LabelledToggleSwitch
                            value={openedToExternalUsers}
                            onChange={onExternalAccessChange}
                            label={_t("Allow external users to join this room")}
                            disabled={disabled || openedToExternalUsers}
                        />
                    </span>
                </div>
            );
        }

        definitions.push({
            value: JoinRule.Invite,
            label: _t("Private (invite only)"),
            description: privateRoomDescription,
            checked: true,
        });
    } else if (joinRule === JoinRule.Public) {
        definitions.push({
            value: JoinRule.Public,
            label: _t("Public"),
            description: (
                <>
                    {_t("Anyone can find and join.")}
                    {aliasWarning}
                </>
            ),
        });
    }

    /**
     * TCHAP : disable space-related options if create space feature is not enabled
     */
    if (TchapUIFeature.isSpaceDisplayEnabled) {
        if (roomSupportsRestricted || preferredRestrictionVersion || joinRule === JoinRule.Restricted) {
            let upgradeRequiredPill;
            if (preferredRestrictionVersion) {
                upgradeRequiredPill = (
                    <span className="mx_JoinRuleSettings_upgradeRequired">{_t("Upgrade required")}</span>
                );
            }

            let description;
            if (joinRule === JoinRule.Restricted && restrictedAllowRoomIds?.length) {
                // only show the first 4 spaces we know about, so that the UI doesn't grow out of proportion there are lots.
                const shownSpaces = restrictedAllowRoomIds
                    .map((roomId) => cli.getRoom(roomId))
                    .filter((room) => room?.isSpaceRoom())
                    .slice(0, 4);

                let moreText;
                if (shownSpaces.length < restrictedAllowRoomIds.length) {
                    if (shownSpaces.length > 0) {
                        moreText = _t("& %(count)s more", {
                            count: restrictedAllowRoomIds.length - shownSpaces.length,
                        });
                    } else {
                        moreText = _t("Currently, %(count)s spaces have access", {
                            count: restrictedAllowRoomIds.length,
                        });
                    }
                }

                const onRestrictedRoomIdsChange = (newAllowRoomIds: string[]) => {
                    if (!arrayHasDiff(restrictedAllowRoomIds || [], newAllowRoomIds)) return;

                    if (!newAllowRoomIds.length) {
                        setContent({
                            join_rule: JoinRule.Invite,
                        });
                        return;
                    }

                    setContent({
                        join_rule: JoinRule.Restricted,
                        allow: newAllowRoomIds.map((roomId) => ({
                            type: RestrictedAllowType.RoomMembership,
                            room_id: roomId,
                        })),
                    });
                };

                const onEditRestrictedClick = async () => {
                    const restrictedAllowRoomIds = await editRestrictedRoomIds();
                    if (!Array.isArray(restrictedAllowRoomIds)) return;
                    if (restrictedAllowRoomIds.length > 0) {
                        onRestrictedRoomIdsChange(restrictedAllowRoomIds);
                    } else {
                        onChange(JoinRule.Invite);
                    }
                };

                description = (
                    <div>
                        <span>
                            {_t(
                                "Anyone in a space can find and join. <a>Edit which spaces can access here.</a>",
                                {},
                                {
                                    a: (sub) => (
                                        <AccessibleButton
                                            disabled={disabled}
                                            onClick={onEditRestrictedClick}
                                            kind="link_inline"
                                        >
                                            {sub}
                                        </AccessibleButton>
                                    ),
                                },
                            )}
                        </span>

                        <div className="mx_JoinRuleSettings_spacesWithAccess">
                            <h4>{_t("Spaces with access")}</h4>
                            {shownSpaces.map((room) => {
                                return (
                                    <span key={room.roomId}>
                                        <RoomAvatar room={room} height={32} width={32} />
                                        {room.name}
                                    </span>
                                );
                            })}
                            {moreText && <span>{moreText}</span>}
                        </div>
                    </div>
                );
            } else if (SpaceStore.instance.activeSpaceRoom) {
                description = _t(
                    "Anyone in <spaceName/> can find and join. You can select other spaces too.",
                    {},
                    {
                        spaceName: () => <b>{SpaceStore.instance.activeSpaceRoom.name}</b>,
                    },
                );
            } else {
                description = _t("Anyone in a space can find and join. You can select multiple spaces.");
            }

            definitions.splice(1, 0, {
                value: JoinRule.Restricted,
                label: (
                    <>
                        {_t("Space members")}
                        {upgradeRequiredPill}
                    </>
                ),
                description,
                // if there are 0 allowed spaces then render it as invite only instead
                checked: joinRule === JoinRule.Restricted && !!restrictedAllowRoomIds?.length,
            });
        }
    }

    const onChange = async (joinRule: JoinRule) => {
        const beforeJoinRule = content.join_rule;

        let restrictedAllowRoomIds: string[];
        if (joinRule === JoinRule.Restricted) {
            if (beforeJoinRule === JoinRule.Restricted || roomSupportsRestricted) {
                // Have the user pick which spaces to allow joins from
                restrictedAllowRoomIds = await editRestrictedRoomIds();
                if (!Array.isArray(restrictedAllowRoomIds)) return;
            } else if (preferredRestrictionVersion) {
                // Block this action on a room upgrade otherwise it'd make their room unjoinable
                const targetVersion = preferredRestrictionVersion;

                let warning: JSX.Element;
                const userId = cli.getUserId();
                const unableToUpdateSomeParents = Array.from(SpaceStore.instance.getKnownParents(room.roomId)).some(
                    (roomId) => !cli.getRoom(roomId)?.currentState.maySendStateEvent(EventType.SpaceChild, userId),
                );
                if (unableToUpdateSomeParents) {
                    warning = (
                        <b>
                            {_t(
                                "This room is in some spaces you're not an admin of. " +
                                    "In those spaces, the old room will still be shown, " +
                                    "but people will be prompted to join the new one.",
                            )}
                        </b>
                    );
                }

                Modal.createDialog(RoomUpgradeWarningDialog, {
                    roomId: room.roomId,
                    targetVersion,
                    description: (
                        <>
                            {_t(
                                "This upgrade will allow members of selected spaces " +
                                    "access to this room without an invite.",
                            )}
                            {warning}
                        </>
                    ),
                    doUpgrade: async (
                        opts: IFinishedOpts,
                        fn: (progressText: string, progress: number, total: number) => void,
                    ): Promise<void> => {
                        const roomId = await upgradeRoom(
                            room,
                            targetVersion,
                            opts.invite,
                            true,
                            true,
                            true,
                            (progress) => {
                                const total = 2 + progress.updateSpacesTotal + progress.inviteUsersTotal;
                                if (!progress.roomUpgraded) {
                                    fn(_t("Upgrading room"), 0, total);
                                } else if (!progress.roomSynced) {
                                    fn(_t("Loading new room"), 1, total);
                                } else if (progress.inviteUsersProgress < progress.inviteUsersTotal) {
                                    fn(
                                        _t("Sending invites... (%(progress)s out of %(count)s)", {
                                            progress: progress.inviteUsersProgress,
                                            count: progress.inviteUsersTotal,
                                        }),
                                        2 + progress.inviteUsersProgress,
                                        total,
                                    );
                                } else if (progress.updateSpacesProgress < progress.updateSpacesTotal) {
                                    fn(
                                        _t("Updating spaces... (%(progress)s out of %(count)s)", {
                                            progress: progress.updateSpacesProgress,
                                            count: progress.updateSpacesTotal,
                                        }),
                                        2 + progress.inviteUsersProgress + progress.updateSpacesProgress,
                                        total,
                                    );
                                }
                            },
                        );
                        closeSettingsFn();

                        // switch to the new room in the background
                        dis.dispatch<ViewRoomPayload>({
                            action: Action.ViewRoom,
                            room_id: roomId,
                            metricsTrigger: undefined, // other
                        });

                        // open new settings on this tab
                        dis.dispatch({
                            action: "open_room_settings",
                            initial_tab_id: ROOM_SECURITY_TAB,
                        });
                    },
                });

                return;
            }

            // when setting to 0 allowed rooms/spaces set to invite only instead as per the note
            if (!restrictedAllowRoomIds.length) {
                joinRule = JoinRule.Invite;
            }
        }

        if (beforeJoinRule === joinRule && !restrictedAllowRoomIds) return;
        if (beforeChange && !(await beforeChange(joinRule))) return;

        const newContent: IJoinRuleEventContent = {
            join_rule: joinRule,
        };

        // pre-set the accepted spaces with the currently viewed one as per the microcopy
        if (joinRule === JoinRule.Restricted) {
            newContent.allow = restrictedAllowRoomIds.map((roomId) => ({
                type: RestrictedAllowType.RoomMembership,
                room_id: roomId,
            }));
        }

        setContent(newContent);
    };

    return (
        <StyledRadioGroup
            name="joinRule"
            value={joinRule}
            onChange={onChange}
            definitions={definitions}
            disabled={disabled}
            className="mx_JoinRuleSettings_radioButton"
        />
    );
};

export default JoinRuleSettings;
