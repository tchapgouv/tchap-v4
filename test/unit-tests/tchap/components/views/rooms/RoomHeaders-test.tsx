import React from "react";
import { KnownMembership, PendingEventOrdering, Room } from "matrix-js-sdk/src/matrix";
import { screen, render, RenderOptions, getByLabelText, queryByLabelText } from "@testing-library/react";

import { mkRoomMember, stubClient } from "~matrix-react-sdk/test/test-utils";
import RoomHeader from "~matrix-react-sdk/src/components/views/rooms/RoomHeader";
import DMRoomMap from "~matrix-react-sdk/src/utils/DMRoomMap";
import { MatrixClientPeg } from "~matrix-react-sdk/src/MatrixClientPeg";
import MatrixClientContext from "~matrix-react-sdk/src/contexts/MatrixClientContext";
import SdkConfig from "~matrix-react-sdk/src/SdkConfig";
import SettingsStore from "~matrix-react-sdk/src/settings/SettingsStore";
import { UIFeature } from "~matrix-react-sdk/src/settings/UIFeature";

function getWrapper(): RenderOptions {
    return {
        wrapper: ({ children }) => (
            <MatrixClientContext.Provider value={MatrixClientPeg.safeGet()}>{children}</MatrixClientContext.Provider>
        ),
    };
}

/**
 *
 * @param count the number of users to create
 */
function mockRoomMembers(room: Room, count: number) {
    const members = Array(count)
        .fill(0)
        .map((_, index) => ({
            userId: `@user-${index}:example.org`,
            name: `Member ${index}`,
            rawDisplayName: `Member ${index}`,
            roomId: room.roomId,
            membership: KnownMembership.Join,
            getAvatarUrl: () => `mxc://avatar.url/user-${index}.png`,
            getMxcAvatarUrl: () => `mxc://avatar.url/user-${index}.png`,
        }));

    room.currentState.setJoinedMemberCount(members.length);
    room.getJoinedMembers = jest.fn().mockReturnValue(members);
}

describe("RoomHeader", () => {
    let room: Room;
    const ROOM_ID = "!1:example.org";
    const featurethreadName: string = "feature_thread";
    const featureVideoName: string = "feature_video_call";
    const featureVideoGroupName: string = "feature_video_group_call";
    const homeserverName: string = "my.home.server";

    const addHomeserverToMockConfig = (homeservers: string[], feature: string) => {
        // mock SdkConfig.get("tchap_features")
        const config: ConfigOptions = { tchap_features: {} };
        config.tchap_features[feature] = homeservers;
        SdkConfig.put(config);
    };

    const getComponent = () => render(<RoomHeader room={room} />, getWrapper());

    function mockDMRoom(memberCount: number = 2) {
        mockRoomMembers(room, memberCount);
        jest.spyOn(SettingsStore, "getValue").mockImplementation(() => false);
        jest.spyOn(room.currentState, "mayClientSendStateEvent").mockReturnValue(false);
        jest.spyOn(room, "getMember").mockReturnValue(mkRoomMember(room.roomId, "@alice:example.org"));

        DMRoomMap.setShared({
            getUserIdForRoomId: () => {
                return "@alice:example.org";
            },
        } as unknown as DMRoomMap);
    }

    beforeEach(async () => {
        const mockClient = stubClient();
        room = new Room(ROOM_ID, MatrixClientPeg.get()!, "@alice:example.org", {
            pendingEventOrdering: PendingEventOrdering.Detached,
        });

        jest.spyOn(mockClient, "getDomain").mockImplementation(() => homeserverName);
        jest.spyOn(room, "isElementVideoRoom").mockReturnValue(false);
        jest.spyOn(room, "isCallRoom").mockReturnValue(false);
        jest.spyOn(room, "isSpaceRoom").mockReturnValue(false);
        jest.spyOn(room, "getType").mockReturnValue(undefined);

        // allow element calls
        jest.spyOn(room.currentState, "mayClientSendStateEvent").mockReturnValue(true);
        // activate the group and widget features
        jest.spyOn(SettingsStore, "getValue").mockImplementation(
            (feature) => feature === "feature_group_calls" || feature == UIFeature.Widgets,
        );

        DMRoomMap.setShared({
            getUserIdForRoomId: jest.fn(),
        } as unknown as DMRoomMap);
    });

    afterEach(() => {
        SdkConfig.reset();
        jest.restoreAllMocks();
    });

    it("should render as expected", async () => {
        const { container } = getComponent();
        expect(container).toMatchSnapshot();
    });

    it("renders the room header", () => {
        const { container } = getComponent();
        expect(container).toHaveTextContent(ROOM_ID);
    });

    it("display well the thread button when feature is activated", () => {
        addHomeserverToMockConfig(["*"], featurethreadName);

        getComponent();

        expect(screen.queryByRole("button", { name: "Threads" })).toBeInTheDocument();
    });

    it("hides the thread button when feature is deactivated", () => {
        addHomeserverToMockConfig(["other.homeserver"], featurethreadName);

        getComponent();

        expect(screen.queryByRole("button", { name: "Threads" })).not.toBeInTheDocument();
    });

    // For 1 to 1 video call
    it("display well the video button when feature is activated for 1v1 call", () => {
        addHomeserverToMockConfig([homeserverName], featureVideoName);
        mockDMRoom();

        const { container } = getComponent();

        expect(queryByLabelText(container, "Video call")).toBeInTheDocument();
    });

    it("hides the video button when feature is deactivated for 1v1 call", () => {
        addHomeserverToMockConfig(["other.homeserver"], featureVideoName);
        mockDMRoom();

        const { container } = getComponent();

        expect(queryByLabelText(container, "Video call")).toBeNull();
    });

    it("hides the video button when feature is activated but is not a direct message room", () => {
        addHomeserverToMockConfig([homeserverName], featureVideoName);

        mockDMRoom(4);

        const { container } = getComponent();

        expect(queryByLabelText(container, "Video call")).toBeNull();
    });

    //  for video group element call button
    it("display well the video group button when feature is activated", () => {
        addHomeserverToMockConfig([homeserverName], featureVideoGroupName);

        mockRoomMembers(room, 4);

        const { container } = getComponent();

        expect(getByLabelText(container, "Video call")).toBeInTheDocument();
    });

    it("hides the video group when feature is deactivated", () => {
        addHomeserverToMockConfig(["other.homeserver"], featureVideoGroupName);

        mockRoomMembers(room, 4);

        const { container } = getComponent();

        expect(queryByLabelText(container, "Video call")).toBeNull();
    });
});
