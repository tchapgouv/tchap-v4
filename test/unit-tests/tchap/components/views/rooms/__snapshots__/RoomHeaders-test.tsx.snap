// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomHeader should render as expected 1`] = `
<div>
  <header
    class="mx_Flex mx_RoomHeader light-panel"
    style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-3x);"
  >
    <div
      class="mx_DecoratedRoomAvatar mx_DecoratedRoomAvatar_cutout"
    >
      <div
        class="mx_DecoratedRoomAvatar_positionedParent"
      >
        <span
          class="_avatar_mcap2_17 mx_BaseAvatar _avatar-imageless_mcap2_61"
          data-color="3"
          data-testid="avatar-img"
          data-type="round"
          role="presentation"
          style="--cpd-avatar-size: 40px;"
        >
          !
        </span>
        <div
          class="mx_DecoratedRoomAvatar_icon mx_DecoratedRoomAvatar_icon_forum"
          tabindex="0"
        />
      </div>
       
    </div>
    <button
      aria-label="Room info"
      class="mx_RoomHeader_infoWrapper"
      tabindex="0"
    >
      <div
        class="mx_Box mx_RoomHeader_info mx_Box--flex"
        style="--mx-box-flex: 1;"
      >
        <div
          aria-level="1"
          class="_typography_yh5dq_162 _font-body-lg-semibold_yh5dq_83 mx_RoomHeader_heading"
          dir="auto"
          role="heading"
        >
          <span
            class="mx_RoomHeader_truncated mx_lineClamp"
          >
            !1:example.org
          </span>
        </div>
      </div>
    </button>
    <div
      class="mx_Flex"
      style="--mx-flex-display: flex; --mx-flex-direction: row; --mx-flex-align: center; --mx-flex-justify: start; --mx-flex-gap: var(--cpd-space-2x);"
    >
      <button
        aria-label="Room info"
        class="_icon-button_bh2qc_17"
        role="button"
        style="--cpd-icon-button-size: 32px;"
        tabindex="0"
      >
        <div
          class="_indicator-icon_133tf_26"
          style="--cpd-icon-button-size: 100%;"
        >
          <div />
        </div>
      </button>
    </div>
    <div
      class="_typography_yh5dq_162 _font-body-sm-medium_yh5dq_50"
    >
      <div
        aria-label="0 members"
        class="mx_AccessibleButton mx_FacePile"
        role="button"
        tabindex="0"
      >
        <div
          class="_stacked-avatars_mcap2_111"
        />
        0
      </div>
    </div>
  </header>
</div>
`;
